[ong-wei-hong] iP

ong-wei-hong opened the pull request:
# GOAT
GOAT stands for
 - greatest
 - of 
 - all
 - time
GOAT has the following commands
1. list
2. todo
3. deadline
4. event
5. mark
6. unmark
7. delete
8. find

A GOAT bot can be initialised as below
```java
public class Main {
        public static void main(String[] args) {
                Duke duke = new Duke();
        }
}
```
We have
 - [X] Todos
 - [X] Deadlines
 - [X] Events
 - [ ] Dragons? :dragon:
 > And then there were Dragons - someone [source](https://www.goodreads.com/series/332001-and-they-found-dragons)

Initialise `Duke` **today**!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ong-wei-hong committed:
completed level 0

ong-wei-hong committed:
completed Level-1

ong-wei-hong committed:
copmleted Level-2

ong-wei-hong committed:
completed Level-3

ong-wei-hong committed:
completed Level-4

ong-wei-hong committed:
completed A-TextUiTesting

ong-wei-hong committed:
completed Level-5

ong-wei-hong committed:
completed Level-6

ong-wei-hong committed:
fixed delete bug

ong-wei-hong committed:
refactored printer

ong-wei-hong committed:
refactored command execution logic

ong-wei-hong committed:
added a command factory, refactored existing code to work

ong-wei-hong committed:
Seperated command parsing logic from CommandFactory
TODO: throw and handle appropriate exception from Task.createTask

ong-wei-hong committed:
Seperated parsing logic from CommandFactory

ong-wei-hong committed:
added save file functionality
TODO: refactor printer
TODO: refactor command

ong-wei-hong committed:
linted, completed Level-7

ong-wei-hong committed:
completed level 8

ong-wei-hong committed:
completed OOP refactor
left with some validations

ong-wei-hong committed:
completed A-MoreOOP

ong-wei-hong committed:
changed Duke from static to instance

ong-wei-hong committed:
moved files to packages

ong-wei-hong committed:
Merge branch 'add-gradle-support'

ong-wei-hong committed:
gradle now works.
TODO: add gradle support for linting and junit

ong-wei-hong committed:
Merge branch 'add-gradle-support'

ong-wei-hong committed:
completed JUnit tests

ong-wei-hong committed:
removed data from commits

ong-wei-hong committed:
completed javadocs

ong-wei-hong committed:
linted

ong-wei-hong committed:
added find command

ong-wei-hong committed:
Merge branch 'master' into branch-A-JavaDoc

ong-wei-hong committed:
added javadocs

ong-wei-hong committed:
linted

ong-wei-hong committed:
merged with master

ong-wei-hong committed:
completed checkstyle

StevenLiudw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,93 @@
+
+public class Duke {
+
+  private final String BOT_NAME = "GOAT";
+  private final String FILEPATH = "data/";
+  private final String SAVEFILE_NAME = "save.txt";
```
Consistent naming conventions are followed. However, it's better to follow the Java naming convention of having constant variables (BOT_NAME, FILEPATH, etc.) as static final

glenngnng replied:
Should FILEPATH be changed to FILE_PATH? Perhaps can use underlines to separate the words in constant naming. Same for SAVE_FILE_NAME 😄

StevenLiudw commented on src/main/java/duke/command/AddTaskCommand.java:
While the code is relatively clear, it could benefit from more in-line comments or JavaDoc comments, especially for the public methods and classes.

glenngnng replied:
Agreed. Javadocs here would be useful.

StevenLiudw commented on src/main/java/duke/helper/DatetimeHelper.java:
```diff
@@ -0,0 +1,22 @@
+
+public class DatetimeHelper {
+
+  public static String inputFormat = "yyyy-MM-dd HHmm";
+  public static String exampleDatetime = "2015-09-15 1300";
```
inputFormat and exampleDatetime should be made final constants since they are not meant to change.

StevenLiudw commented on src/main/java/duke/io/FileIO.java:
```diff
@@ -0,0 +1,50 @@
+    this.description = description;
+  }
+
+  public String read() throws FileNotFoundException {
+    File f = new File(filepath + filename);
+    Scanner in = new Scanner(f);
+    StringBuilder sb = new StringBuilder();
+    while (in.hasNext()) {
+      sb.append(in.nextLine().trim());
+      sb.append('\n');
+    }
+    return sb.toString().trim();
+  }
```
inputFormat and exampleDatetime should be made final constants since they are not meant to change.

StevenLiudw commented on src/main/java/duke/io/Printer.java:
```diff
@@ -0,0 +1,60 @@
+
+  public void flush() {
+    String line = "  ____________________________________________________________";
+    System.out.println(line);
+    error.ifPresentOrElse(
+        (s) -> {
+          System.out.println(String.format("    %s", s));
+        },
+        () -> {
+          messages.forEach(
+              (s) -> {
+                System.out.printf("    %s\n", s);
+              });
+        });
+    System.out.println(line);
```
Rather than checking if the error is present using ifPresentOrElse, it might be clearer to check error.isPresent() and then act accordingly.

StevenLiudw commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    if (to.isEmpty()) {
+      throw new DukeException("The to of a event cannot be empty.");
```
Instead of "a event", use "an event".

StevenLiudw commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,69 @@
+
+  public String getNumberOfTasks() {
+    return String.format(
+        "Now you have %d task%s in the list.", tasks.size(), tasks.size() == 1 ? "" : "s");
```
For clarity and consistency, consider renaming the method getNumberOfTasks() to something like getTaskCountMessage().

StevenLiudw reviewed (commented):
LGTM. Quite impressed by the nice OOP structure(divide exception, helper classes and IO). But it is clear that you need to provide more JavaDoc for better user understanding. :)

ong-wei-hong committed:
added gui for duke

glenngnng commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,61 @@
+  protected TaskList taskList;
+  private FileIO savefile;
+
+  private boolean saveAfterAction = false;
```
Perhaps it would be better to change the name to include a prefix like is? A possible change could be isSavedAfterAction. This could be easier to understand that it is a boolean.

glenngnng commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,31 @@
+  public void action() {
+    Task task;
+    try {
+      int i = Integer.parseInt(s); // 1-indexed
```
Not sure if this is OK under the Java coding standard, but perhaps can indent the comment under this line.

glenngnng commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+
+  @Override
+  public boolean filter(Optional<LocalDateTime> before) {
+    return before.filter((datetime) -> datetime.isBefore(by)).isEmpty(); // if is empty return true
```
Same issue with commenting here. Perhaps can push the comment to the next line and indent it the same way. Improves readability if the line is too long.

glenngnng commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,60 @@
+        super.toString(), DatetimeHelper.format(from), DatetimeHelper.format(to));
+  }
+
+  @Override
+  public String toCommand() {
+    return String.format(
+        "%s /from %s /to %s",
+        super.toCommand(), DatetimeHelper.commandFormat(from), DatetimeHelper.commandFormat(to));
+  }
+
+  @Override
+  public boolean filter(Optional<LocalDateTime> before) {
+    return before
+        .filter((datetime) -> datetime.isBefore(from))
+        .isEmpty(); // if is empty return true
+  }
```
Perhaps some javadocs explaining these methods would be useful?

glenngnng commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,69 @@
+    }
+  }
+
+  public static Task createTask(String type, String name, Map<String, String> arguments) {
```
I like this naming, it is very clear! 👍

ong-wei-hong committed:
completed assertions

ong-wei-hong committed:
improved code quality

ong-wei-hong committed:
used streams

ong-wei-hong committed:
completed CI

ong-wei-hong committed:
Merge pull request #1 from ong-wei-hong/branch-A-Streams

Branch a streams

ong-wei-hong committed:
Merge branch 'master' into branch-A-Assertions

ong-wei-hong committed:
Merge pull request #2 from ong-wei-hong/branch-A-Assertions

Branch a assertions

ong-wei-hong committed:
Merge branch 'master' into branch-A-CodeQuality

ong-wei-hong committed:
added assertions for code quality changes

ong-wei-hong committed:
Merge pull request #3 from ong-wei-hong/branch-A-CodeQuality

Branch a code quality

ong-wei-hong committed:
Merge branch 'master' into branch-A-CI

ong-wei-hong committed:
added snooze command

snooze command takes in the index of task to edit and arguments to
amend the attributes

cannot snooze a todo

snoozing a deadline requires a by argument

snoozing an event requires a from or to argument

ong-wei-hong committed:
updated input.txt to provide testing

ong-wei-hong committed:
completed docs

ong-wei-hong committed:
added features in docs

ong-wei-hong committed:
fixed junit tests

ong-wei-hong committed:
checkStyle

ong-wei-hong committed:
addressed issues

ong-wei-hong committed:
moved UI to correct place

ong-wei-hong committed:
moved Ui.png again

ong-wei-hong committed:
removed test for now, to see if github pages is fixed

ong-wei-hong committed:
updated build.gradle
